### DISCUSSION TITLE
Accessing subscribed topics on socket close event

### DISCUSSION THREAD

**Discussion by tayler-king** [uNetworking/uWebSockets.js#869] [Q&A] [answered]
I've previously been tracking topics that individual sockets subscribe to manually, creating a list for each socket.

I felt that this was unnecessary seeing as though you can call `socket.getTopics()` to retrieve a list of subscribed topics.

On the socket `close` event I'd like to do some clean-up, namely retrieving the list of topics the disconnected socket has ben subscribed to. An example of this would be:

```js
uWS.App().ws('/socket', {
  close: socket => emitter.emit('closed', socket)
});

emitter.on('closed', socket => {
  const topics = socket.getTopics();

  topics.forEach(topic => {
    // do something here
  });
});
```

Unfortunately when attempting to call `socket.getTopics()` in the above scenario, I'm experiencing the error `Invalid access of closed uWS.WebSocket/SSLWebSocket.`

Is calling `socket.getTopics()` after the close event fires undefined behaviour, or is this an issue due to how the event loop is handling the emitter callback? I'm still able to access socket data in the emitter's event callback, just not call `getTopics`.

Documentation states _"You may not use WebSocket after this event."_, so I'm lead to believe that the above would be undefined behaviour. In that case, is there a better solution to what I'm wanting to implement other than keeping track of subscribed topics manually?

**Comment by e3dio** [comment]
```javascript
uWS.App().ws('/socket', {
  close: socket => {
    socket.topics = socket.getTopics();
    emitter.emit('closed', socket);
  }
});

emitter.on('closed', socket => {
  socket.topics.forEach(topic => {
    // do something here
  });
});
```

**Comment by uNetworkingAB** [comment, marked as answer]
See https://github.com/uNetworking/uWebSockets/releases/tag/v20.32.0

**Reply by tayler-king** [reply]
Thanks. Seems like the sensible way to do it.
